### PR TITLE
Fix applying special char on translations

### LINKED ISSUE #29940: BO/FO- HTML tags are not well displayed in different pages
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

In the BO - Customer Settings - Groups page, the info alert is not well displayed
image

In the BO - Notifications bell > Orders
Screenshot 2022-12-20 at 14 31 40

In the FO > Create account page, the "Newsletter subscription" block and the block "Customer data privacy" are not well displayed
image

> Edited by @HanaRebaiQA 

When enable multistore and add a new shop, the link tag is bad displayed!
image

### Steps to reproduce

**Case1**
1. Go to BO - Customer Settings - Groups page
2. See error
3. Go to FO 
4. Click on Sign in
5. Create an account 
6. See error 

PS8.0.0RC1 => oK

> Edited by @HanaRebaiQA 

**Case2**
1. Go to BO > General > Enable Multistore
2. In BO > Advanced parameters > Multistore > Add a new shop
3. In the shop name > see the help message

>Edited by @khouloudbelguith 

**Case3**
1. Go to BO > Advanced Parameters > Performance page
2. Disable Features and save
3. Go to BO > Catalog > Attributes & Features > Features page

image

**Case4**
In the FO > Create account page, the "Newsletter subscription" block and the block "Customer data privacy" are not well displayed
image

>Edited by Khouloud

**Case5**

1. Go to BO > Advanced Parameters > Import page
2. Try to import a valid file csv, for example in my case products.csv
3. Click on Import > A modal "Importing data" is displayed
4. See error `<span id="validate_progression_done">0</span>% validated`

image

### Expected behavior

image
It should be 
image

### PrestaShop version(s) where the bug happened

8.0.x, develop

### PHP version(s) where the bug happened

7.2, 7.3

### If your bug is related to a module, specify its name and its version

_No response_

### LINKED ISSUE #29959: BO - Import - Popup for data import has non-escaped text
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

When I do an import, at the end, there's a popup.
This popup should not have non escaped text:
Screenshot 2022-10-11 at 14 18 05

It's a regression, it 1.7.8.7, it looked normal:
Screenshot 2022-10-11 at 15 21 19

### Expected behavior

I should see the text without the HTML elements.

### Steps to reproduce

1. Go to BO > Advanced Parameters > Import
2. Download import file of your choice
3. Choose your file to import > match all fields > Import
4. See the popup appear at the end

### PrestaShop version(s) where the bug happened

8.0.x, develop

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

_No response_

### LINKED ISSUE #30539: Delete _raw parameter in trans methods
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Automatically escaping HTML characters in translation methods is not a good practice because it produces problems like #29959. 

A temporary workaround has been added in https://github.com/PrestaShop/PrestaShop/pull/30415 but it's not a real fix.

### Describe the solution you'd like

The html escaping code should be removed alongside the temporary workaround, to make it so the trans methods no longer escape HTML on their own.

All calls to translation methods which relied on HTML escaping should be adapted in one of these ways:

```php
// escape everything
htmlspecialchars(
	$this->trans('A string with %replacement%', ['%replacement%' => $codeInjectionGoesHere], 'Domain.domain.domain')
);

// escape only the injection
$this->trans(
    'A string with <b>HTML that we want to keep</b> and %replacement%',
    ['%replacement%' => htmlspecialchars($codeInjectionGoesHere)],
    'Domain.domain.domain'
);
```

This can continue to be automatic in twig and smarty, where dedicated methods exist to control whether the result is escaped or not.

### Alternatives you've considered

_No response_

### Additional context

Can only be developed when https://github.com/PrestaShop/PrestaShop/pull/30415 is merged

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [x] I'm willing to submit a Pull Request that implements this feature.
- [x] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by mflasquin**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | Move translation sanitize before applying SF translation to avoid sanitize on html in trans parameters
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Resolves #29959, resolves #29940 
| Related PRs       | [#29584](https://github.com/PrestaShop/PrestaShop/pull/29584)
| How to test?      | See #29959 #29940 
| Possible impacts? | Check also there is no regression on https://github.com/PrestaShop/PrestaShop/issues/28877

Escape html parameter in translator is not a good practice. This task must be done before to send it to the translator. To do not introduce a bc break with deletion of applying htmlspecialchars in trans method, we introduce the _raw parameter that can be sent to not apply htmlspecialchars on the translation. This parameter will be removed in the next major version.

### CODE REVIEW

**Review by yanmakouf — DISMISSED** [review]
Maybe the failing test is wrong

**Review by zuk3975 — DISMISSED** [review]

**Review by nicosomb — COMMENTED** [review]
It seems ok for me but did you see this discussion https://github.com/PrestaShop/PrestaShop/issues/29940#issuecomment-1274251774 ? Just to be sure that all is ok with this fix.

**Comment by mflasquin** [comment]
> It seems ok for me but did you see this discussion [#29940 (comment)](https://github.com/PrestaShop/PrestaShop/issues/29940#issuecomment-1274251774) ? Just to be sure that all is ok with this fix.

Yes but I wonder if my fix can solve the problem. This is why I've ping @eternoendless to review this PR

File: tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php
Comment (eternoendless) [inline tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php]: htmlspecialchars should be escaping the content of the replaced token

**Review by eternoendless — CHANGES_REQUESTED** [review]
This has been fixed and unfixed, then fixed again, then unfixed again - several times.

In twig, you'd do `{{ trans('Foo')|raw }}` if you wanted to use HTML inside the translated content. Logically, smarty should behave the same.

I think there's a design decision to be made about either accepting that you can't use HTML inside translated tokens on Smarty, or removing the use `htmlspecialchars` at the translator level altogether (and applying it at templating level, like twig does)

This change is opening up the system to code injection attacks

File: tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php
Comment (mflasquin) [reply tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php]: So we have to replace all trans tag to add 'raw' each place we want to use HTML in the BO ?

Comment (eternoendless) [reply tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php]: This is already necessary in twig templates because twig escapes HTML by default

Screenshot 2022-11-28 at 17 01 51

(Source: https://symfony.com/doc/4.4/translation.html#using-twig-filters)

Smarty templates seem to do this by default in [admin controllers](https://github.com/PrestaShop/PrestaShop/blob/8.0.0/config/smartyadmin.config.inc.php#L75) but apparently not on [front controllers](https://github.com/PrestaShop/PrestaShop/blob/8.0.0/config/smartyfront.config.inc.php#L160).

There's clearly some homogeneization to do.

**Comment by atomiix** [comment]
@mflasquin friendly reminder :)

File: tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php
Comment (mflasquin) [reply tests/Integration/PrestaShopBundle/TranslationIntegrationTest.php]: @eternoendless changed as discussed yesterday

File: classes/ObjectModel.php
Comment (eternoendless) [inline classes/ObjectModel.php]: Why did you make this public?

File: classes/ObjectModel.php
Comment (eternoendless) [inline classes/ObjectModel.php]: The deprecation and unset should be done only if the parameter is present.
The deprecation message could be improved as well.

```suggestion
        if (isset($parameters['_raw'])) {
            @trigger_error(
                'The _raw parameter is deprecated and will be removed in the next major version.',
                E_USER_DEPRECATED
            );
            unset($parameters['_raw']);
            
            return $this->translator->trans($id, $parameters, $domain, $locale);
        }
        
        return htmlspecialchars($this->translator->trans($id, $parameters, $domain, $locale), ENT_NOQUOTES);
```

File: classes/controller/AdminController.php
Comment (eternoendless) [inline classes/controller/AdminController.php]: You're removing the replacement tokens here, you need to keep them.
The problem can be solved more easily in this function because it already had a dedicated $htmlentities parameter

```php
if ($htmlentities === true) {
    return htmlspecialchars($this->translator->trans($string, $parameters), ENT_NOQUOTES);
}

return $this->translator->trans($string, $parameters);
```

File: classes/controller/Controller.php
Comment (eternoendless) [inline classes/controller/Controller.php]: Same change as in the other function:

```suggestion
        if (isset($parameters['_raw'])) {
            @trigger_error(
                'The _raw parameter is deprecated and will be removed in the next major version.',
                E_USER_DEPRECATED
            );
            unset($parameters['_raw']);
            
            return $this->translator->trans($id, $parameters, $domain, $locale);
        }
        
        return htmlspecialchars($this->translator->trans($id, $parameters, $domain, $locale), ENT_NOQUOTES);
```

File: classes/module/Module.php
Comment (eternoendless) [inline classes/module/Module.php]: Same change as in the other function:

```suggestion
        if (isset($parameters['_raw'])) {
            @trigger_error(
                'The _raw parameter is deprecated and will be removed in the next major version.',
                E_USER_DEPRECATED
            );
            unset($parameters['_raw']);
            
            return $this->translator->trans($id, $parameters, $domain, $locale);
        }
        
        return htmlspecialchars($this->translator->trans($id, $parameters, $domain, $locale), ENT_NOQUOTES);
```

File: src/PrestaShopBundle/Translation/PrestaShopTranslatorTrait.php
Comment (eternoendless) [inline src/PrestaShopBundle/Translation/PrestaShopTranslatorTrait.php:57]: We could add a deprecation message here as well, warning that it's no longer supported

File: tests/Integration/Classes/ObjectModelTest.php
Comment (eternoendless) [inline tests/Integration/Classes/ObjectModelTest.php:105]: I consider a good practice to describe the purpose of this test. Without context, it's very difficult to understand what you are testing here.

**Review by eternoendless — CHANGES_REQUESTED** [review]

**Comment by eternoendless** [comment]
Please, for the record, explain why we chose to go with _raw and create an issue to remove it in 9.0 so that it's not forgotten

File: classes/ObjectModel.php
Comment (mflasquin) [reply classes/ObjectModel.php]: It's a mistake

File: classes/controller/AdminController.php
Comment (mflasquin) [reply classes/controller/AdminController.php]: $parameters was already defined as an empty array in the function before changes

**Comment by mflasquin** [comment]
> Please, for the record, explain why we chose to go with _raw and create an issue to remove it in 9.0 so that it's not forgotten

I just add more details in the PR 
And create an issue : https://github.com/PrestaShop/PrestaShop/issues/30539

File: tests/Integration/Classes/ObjectModelTest.php
Comment (eternoendless) [inline tests/Integration/Classes/ObjectModelTest.php]:
```suggestion
     * Check if html in trans is not escaped when the _raw parameter is used
```

File: tests/Integration/Classes/controller/AdminControllerTest.php
Comment (eternoendless) [inline tests/Integration/Classes/controller/AdminControllerTest.php]:
```suggestion
     * Check if html in trans is not escaped when the _raw parameter is used
```

File: tests/Integration/Classes/module/ModuleTest.php
Comment (eternoendless) [inline tests/Integration/Classes/module/ModuleTest.php]:
```suggestion
     * Check if html in trans is escaped when the _raw parameter is used
```

**Review by eternoendless — COMMENTED** [review]
Final comments

**Review by eternoendless — DISMISSED** [review]

File: controllers/admin/AdminGroupsController.php
Comment (atomiix) [inline controllers/admin/AdminGroupsController.php:649]: If using `_raw` is deprecated, maybe we should not use it there?

Comment (mflasquin) [reply controllers/admin/AdminGroupsController.php:649]: Because for the moment we have to keep the htmlentities on trans method to avoid BC break. In the next major version we will delete the call to htmlentities in trans method and we will have to make it outside.
On this example we want to keep html, so we have to precise "_raw" to do not apply htmlentities

**Review by nicosomb — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @mflasquin ,

I have launched the automated tests: ✅ https://github.com/florine2623/testing_pr/actions/runs/3674412750

I still have some non-escaped content here:

In FO: :x:
Screenshot 2022-12-12 at 11 39 20

In BO > Multistore: :x: 
Screenshot 2022-12-12 at 11 43 45

In BO > Attributes & Features > Features tab: :x:
Screenshot 2022-12-12 at 11 45 38

On **Import** and **Customer Group** page, it is ✅ 

And there's no regression on https://github.com/PrestaShop/PrestaShop/issues/28877: ✅ 
Screenshot 2022-12-12 at 11 53 25
Screenshot 2022-12-12 at 11 53 44


Could you check?
Thanks!

**Comment by mflasquin** [comment]
> Hello @mflasquin ,
> 
> I have launched the automated tests: white_check_mark https://github.com/florine2623/testing_pr/actions/runs/3674412750
> 
> I still have some non-escaped content here:
> 
> In FO: x Screenshot 2022-12-12 at 11 39 20
> 
> In BO > Multistore: x Screenshot 2022-12-12 at 11 43 45
> 
> In BO > Attributes & Features > Features tab: x Screenshot 2022-12-12 at 11 45 38
> 
> On **Import** and **Customer Group** page, it is white_check_mark
> 
> And there's no regression on #28877: white_check_mark Screenshot 2022-12-12 at 11 53 25 Screenshot 2022-12-12 at 11 53 44
> 
> Could you check? Thanks!

Issues on FO must be fixed on dedicated modules (ps_dataprivacy and ps_emailsubscription) => FYI @eternoendless

**Review by eternoendless — DISMISSED** [review]

**Comment by eternoendless** [comment]
@mflasquin please consider fixing the related modules

**Comment by kpodemski** [comment]
It seems that this change will break many modules in the ecosystem.

**Comment by mflasquin** [comment]
> It seems that this change will break many modules in the ecosystem.

Not for the moment, issue with not interpreted html already exist, we give a way to fix it when precise '_raw' parameter to do not apply htmlspecialchars on translated string. But in 9.0 we will avoid applying htmlspecialchars in trans method, this will be a BC break.

**Review by zuk3975 — DISMISSED** [review]

File: controllers/admin/AdminLoginController.php
Comment (FabienPapet) [inline controllers/admin/AdminLoginController.php:94]: If I understand correctly, you are adding a deprecated parameters, so this will break in the next version

Comment (mflasquin) [reply controllers/admin/AdminLoginController.php:94]: This will not break but will be not necessary. On the other hand you will have to escape parameters before send them to the translation method. And we will have to communicate to prevent this behaviour to avoid security flaws.

**Review by sallemiines — COMMENTED** [review]
Hello @mflasquin 

Thanks for the PR !

Unfortunately the pr is not validate ! 
check the different Fixed ticket and here is the report --


1. Issue [#29959](https://github.com/PrestaShop/PrestaShop/issues/29959) is well fixed  :heavy_check_mark: 

image

2. issue [#29940 ](https://github.com/PrestaShop/PrestaShop/issues/29940): 

**case 1 :**  first error is fixed :heavy_check_mark: 

image

**Second error is NOK**  :x: 

image

**case 2** : the error is well fixed :heavy_check_mark: 

image

**case 3** : the error is well fixed :heavy_check_mark: 

image

**case 4**  : the error is well fixed :heavy_check_mark: 

image

Please check and feedback ! 
Thank you !

**Comment by mflasquin** [comment]
> Hello @mflasquin
> 
> Thanks for the PR !
> 
> Unfortunately the pr is not validate ! check the different Fixed ticket and here is the report --
> 
> 1. Issue [#29959](https://github.com/PrestaShop/PrestaShop/issues/29959) is well fixed  heavy_check_mark
> 
> image
> 
> 2. issue [#29940 ](https://github.com/PrestaShop/PrestaShop/issues/29940):
> 
> **case 1 :** first error is fixed heavy_check_mark
> 
> image
> 
> **Second error is NOK** x
> 
> image
> 
> **case 2** : the error is well fixed heavy_check_mark
> 
> image
> 
> **case 3** : the error is well fixed heavy_check_mark
> 
> image
> 
> **case 4** : the error is well fixed heavy_check_mark
> 
> image
> 
> Please check and feedback ! Thank you !

@sallemiines the error should be fixed in ps_dataprivacy and ps_emailsubscription when this PR will be merged

https://github.com/PrestaShop/ps_emailsubscription/pull/91
https://github.com/PrestaShop/ps_dataprivacy/pull/40

**Comment by MhiriFaten** [comment]
Hello @mflasquin ,

Did you mean that the error of this PR will be fixed only when those 2 PRs https://github.com/PrestaShop/ps_emailsubscription/pull/91 and https://github.com/PrestaShop/ps_dataprivacy/pull/40 will be merged??

If that is what you mean so I will add a '**blocked**' label to this PR until the other PRs will be merged.

Waiting for your feedback :pray: !

Thank you.

**Comment by HanaRebaiQA** [comment]
Hello @mflasquin 

This PR is supposed to fix the issue that was communicated above (Case 4 of[ #29940](https://github.com/PrestaShop/PrestaShop/issues/29940))
If I understood correctly, these two PRs of modules [#ps_emailsubscription/pull/91](https://github.com/PrestaShop/ps_emailsubscription/pull/91) and [#ps_dataprivacy/pull/40](https://github.com/PrestaShop/ps_dataprivacy/pull/40) will fix the error. 
In this case, normally we should check the 3 PRs ([The current one](https://github.com/PrestaShop/PrestaShop/pull/30415), [#ps_emailsubscription/pull/91](https://github.com/PrestaShop/ps_emailsubscription/pull/91) and [#ps_dataprivacy/pull/40](https://github.com/PrestaShop/ps_dataprivacy/pull/40) ) at the same time and together for the issue to be totally fixed. 

That's right? What do you think ?

Thank You!

**Comment by eternoendless** [comment]
IMO this PR shouldn't be blocked by the PRs in modules. The bugs in modules depend on this PR and not the other way around. So to me, this PR is valid and should be merged regardless of what happens with the other PRs.

**Comment by mflasquin** [comment]
> IMO this PR shouldn't be blocked by the PRs in modules. The bugs in modules depend on this PR and not the other way around. So to me, this PR is valid and should be merged regardless of what happens with the other PRs.

@eternoendless what do yo think about recreate this PR targetting 8.0.x branch ?

**Comment by eternoendless** [comment]
Sounds good to me

**Comment by mflasquin** [comment]
FYI I have change the target branch to 8.0.x

**Review by eternoendless — APPROVED** [review]

**Review by 0x346e3730 — APPROVED** [review]

**Comment by FabienPapet** [comment]
@jolelievre @atomiix @matthieu-rolland can one of you please review this ? I'm not sure to fully understand this PR and its impacts for the project

**Comment by Hlavtox** [comment]
Guys, I don't like these things at all. If I understand correctly, does it mean that all translations with html tags are broken on v8 unless they add _raw parameter like here? 🤔 https://github.com/PrestaShop/ps_emailsubscription/pull/91/files

I hope I am missing something. 😛

**Comment by eternoendless** [comment]
> If I understand correctly, does it mean that all translations with html tags are broken on v8 unless they add _raw parameter like here?

@Hlavtox no, not all. The ones impacted by this change are those being used with `Controller::trans()`. This problem exists since 1.7.0.2 because since then translations are sent through `htmlspecialchars` every time `Controller::trans()` is called **with parameters** ([see here](https://github.com/PrestaShop/PrestaShop/pull/7026)). This should have never been done automatically like so, but fixing it will be a BC, and we can't do that until v9... which we will.

File: tests/Integration/Classes/ObjectModelTest.php
Comment (jolelievre) [inline tests/Integration/Classes/ObjectModelTest.php:108]: Since `TestableObjectModel` is a class purely dedicated to testing the ObjectModel class, you could add a public method as a bridge to the protected `trans` method without impact. Even with a naming as lousy as `public function testTrans`
It would avoid "cheating" with this reflection to make the protected method accessible ^^

Although it's just a suggestion, not worth blocking the PR

Comment (jolelievre) [reply tests/Integration/Classes/ObjectModelTest.php:108]: However, add a use statement for `ReflectionMethod` would be nice

File: tests/Integration/Classes/controller/AdminControllerTest.php
Comment (jolelievre) [inline tests/Integration/Classes/controller/AdminControllerTest.php:102]: Hmm since you test it here the same way as in ObjectModel, my initial comment about a bridge method makes less sense At least if you use this everywhere there's some consistency in the code

**Review by jolelievre — APPROVED** [review]
Thanks @mflasquin 

Just a suggestion to add use statements, but not blocking

**Review by matks — APPROVED** [review]

**Comment by paulnoelcholot** [comment]
Hello @mflasquin!
I have tested your PR and here is my report:

CASE 1 == OK to 8.0.x and develop
CASE 2 == OK to 8.0.x and develop
CASE 3 == OK to 8.0.x and develop

CASE 4 == NOK to 8.0.x and develop
image

CASE 5 == OK to 8.0.x and develop but I have seen another issue
image

**Comment by mflasquin** [comment]
@paulnoelcholot It's normal for the case 4, we need to merge this PR and then the issue will be fixed in https://github.com/PrestaShop/ps_emailsubscription/pull/91
https://github.com/PrestaShop/ps_dataprivacy/pull/40

**Comment by Hlavtox** [comment]
@paulnoelcholot Case 5 is not an issue no? It's just import data.

**Comment by eternoendless** [comment]
Case 5 looks like normal behavior to me, you don't want to display imported HTML there.

**Comment by paulnoelcholot** [comment]
Thank you @Hlavtox and @eternoendless for your comments. So it's ok for me! :tada:

**Review by paulnoelcholot — APPROVED** [review]

**Comment by mflasquin** [comment]
Thanks you all 
@eternoendless @paulnoelcholot

**Comment by SharakPL** [comment]
I've just installed PS 8.0.2 (classic theme, all modules upgraded) and it still doesn't look fixed

obraz

Probably because I have ps_emailsubscription at v.2.7.0. Fix is [in 2.7.1](https://github.com/PrestaShop/ps_emailsubscription/pull/91) but even though it's been released in January the update is still not available 😢 

obraz

It's the same with ps_dataprivacy. v.2.1.1 update not available from BO even with ps_mbo installed. [Have to install uzing zip](https://github.com/PrestaShop/ps_dataprivacy/releases)

___
I can confirm that after manual module upgrades it's all fine on FO.

**Comment by mflasquin** [comment]
Can you try to install ps_distributionapiclient, I think it will purpose the update

**Comment by SharakPL** [comment]
I have this module. Problem is my shop is developed on localhost and it looks like API refuses connections from 127.0.0.1. Live shops update all modules without problems.

Besides, like I mentioned before, manual upgrade of modules fixed it for me so it's all good 👍

**Pull request merged**
